Integrate fpc-validation as compiled validator

mbrandenburger opened the pull request:
**What this PR does / why we need it**:

This PR integrates FPC custom validation into the peer without using go plugins. 

- Add FPC Validators (including ERCC and ERCC-decorator) to handler library using FPC build tag
- move from plugin folder to validation folder
- Rename ecc-vscc to fpc-vscc
- Deprecate ercc-vscc and ercc decorator
- Align FPCValdation plugin interface with DefaultValdator
- use protoutils to unmarshal all protos
- refactor checkEnclaveEndorsement for better readability and
maintainability

**Which issue(s) this PR fixes**:
<!--
  list existing bug, feature and/or work-item which this PR addresses.
  You might also consider creating an issue first for the PR.
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing changes and/or breaks backward compatability?**:
<!--
  If no, you can delete this section
  If yes, describe what changes and/or what breaks ..
-->
```

mbrandenburger commented on fabric/Makefile:
```diff
@@ -75,24 +76,28 @@ clean-patch:
 $(FABRIC_TLCC):
 	ln -sfn $$(/bin/pwd)/../tlcc $(FABRIC_TLCC)
 
+$(FABRIC_VALIDATION):
+	ln -sfn $$(/bin/pwd)/../validation/fpc $(FABRIC_VALIDATION)
+
```
I am wondering if we really need this.

g2flyer replied:
I hope not. that would kindof defeat a bit the purpose of build tags?

bvavala replied:
I did the same... and for a reason.
The validator has to access the Go protobufs, which "we" have in the go path, while "Fabric" has then in the vendor folder.
So, without this link, go build complains that it wants the protobufs in Fabric's vendor folder (following a similar path: `github.com/hyperledger/...`). With this link, instead, go build grabs the protobufs from vendor and it's happy. In this case, we might think of dropping our FPC dependency on protobufs (though TLCC might still require them).

g2flyer replied:
i doubt that the symlinks will be acceptable: what would happen if we also vendor protos in fpc?  That said, wouldn't vendoring become obsolete once you move to go modules. Actually, fabric-protos-go _is_ modularized and fabric does include fabric-protos-go in go.mod. I think the version in vendor is (should be) obsolete!

mbrandenburger replied:
as discussed with @g2flyer we keep the symlinks until we bump to fabric v2.2

g2flyer commented:
This still relies on the plugin framework and so doesn't really help getting rid of plugins and will also require changes once fabric does get rid of plugins.  So imho technically moderately meaningful. But if that puts us Gary off the back, so be it.

From looking at `evaluateModeAndLoad` in `.../library/registry.go`, this should work also for all four plugin types, , so we should do conversion of our other two plugins at the same time.  We should definitely get rid of the symlinks (see separate discussion above).

g2flyer commented on validation/fpc/fpc_validation.go:
```diff
@@ -0,0 +1,114 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
```
Moving stuff out of plugins makes sense if we do not use plugins anymore. But a `git mv` would have kept the changes smaller and kept provenance?

g2flyer replied:
btw: you also will have to update core-fpc.yaml in utils/docker-compose ...

mbrandenburger replied:
updated

g2flyer commented on validation/fpc/fpc_validation_logic.go:
```diff
@@ -0,0 +1,340 @@
+	}
+
+	// validate the payload type
+	if common.HeaderType(chdr.Type) != common.HeaderType_ENDORSER_TRANSACTION {
+		logger.Errorf("Only Endorser Transactions are supported, provided type %d", chdr.Type)
+		err = fmt.Errorf("Only Endorser Transactions are supported, provided type %d", chdr.Type)
+		return nil, err
+	}
```
This is something we haven't done before. From `vendor/github.com/hyperledger/fabric-protos-go/common/common.pb.go` no other type seems relevant (and that's why we probably always received only endorser transactions) but definitely robust programming to explicitly check ...

mbrandenburger replied:
this check we inherited from also from the dafaultValidator https://github.com/hyperledger/fabric/blob/master/core/handlers/validation/builtin/v20/validation_logic.go#L109

mbrandenburger commented on Makefile:
```diff
@@ -6,10 +6,10 @@
 
 FPC_SDK_DEP_DIRS = utils/fabric common ecc_enclave ecc
-FPC_PEER_DEP_DIRS = plugins tlcc_enclave tlcc fabric ecc_enclave ecc # ecc & ecc_enclave we need if we run on host in external builder, not via docker
+FPC_PEER_DEP_DIRS = ercc tlcc_enclave tlcc fabric ecc_enclave ecc # ecc & ecc_enclave we need if we run on host in external builder, not via docker
```
the validation does not have any build artefacts anymore as the code is used when build in the peer in `fabric/`. In contrast, in `plugins/` we compiled the actual plugins, so I would say from regarding build structure they are now equivalent. 

However, since the ercc-vscc needs the `provider.go` in `ercc/attestation/`, we need to compile this first, before running `make` in `fabric/` to build the peer.

mbrandenburger replied:
With ercc refactoring and removing ercc-vscc this build dep will go away.

g2flyer replied:
key phrase, though, "wll", i.e., "did not yet" ;-)

mbrandenburger commented on README.md:
```diff
@@ -376,9 +376,6 @@ enclave and [tlcc/](tlcc) for the ledger enclave.
 In order to run and deploy a chaincode enclave we need to build the enclave
 registry. See [ercc/](ercc).
 
-Both, ecc and ercc use custom validation plugins to verify transactions. In [plugins/](plugins) you can find
```
well, this is the building instructions, as we don't have a build target for the plugins anymore, I thought we could remove it. 

I brought this section back and linked to fabric folder

mbrandenburger commented on validation/fpc/fpc_validation_logic.go:
```diff
@@ -0,0 +1,340 @@
+		return nil, err
+	}
+
+	// ...and the transaction...
+	tx, err := protoutil.UnmarshalTransaction(payl.Data)
+	if err != nil {
+		logger.Errorf("VSCC error: GetTransaction failed, err %s", err)
+		return nil, err
+	}
+
```
I see that the comment read nicely before. Still, I would to keep the order like that as it follows the the implementation of the default validator. https://github.com/hyperledger/fabric/blob/master/core/handlers/validation/builtin/v20/validation_logic.go#L109

g2flyer replied:
Actually, given that you kept it closely (identical) to that code, would have been good to add a comment with the provenance of your code.  I think this is good practice and we definitely will have to do that for tlcc ... (i'm trying to do that for all the config files and docker-compose stuff where applicable: it really helps when doing the upgrades ...)

mbrandenburger commented:
@g2flyer Thanks for the comments, I added back the comment in `README.md`. For other feedback see inline review comments. I squashed and ready to merge (once travis made it through).
